tests/test_analyzer.py

```diff
@@ -0,0 +1,151 @@
+
+from google.protobuf import json_format
+from tritonclient.grpc import model_config_pb2
+
+from .common import test_result_collector as trc
+
+from .mocks.mock_config import MockConfig
+
+
+class TestAnalyzer(trc.TestResultCollector):
+    """
+    Tests the methods of the Analyzer class
+    """
+
+    def mock_get_state_variable(self, name):
+        return {
+            'model1': {
+                'config1': None,
+                'config2': None,
+                'config3': None,
+                'config4': None
+            }
+        }
+
+    @patch(
+        'model_analyzer.state.analyzer_state_manager.AnalyzerStateManager.get_state_variable',
+        mock_get_state_variable)
+    def test_get_analyze_command_help_string(self):
+        """
+        Tests that the member function returning the analyze command help string
+        works correctly.
+        """
+
+        args = [
+            'model-analyzer', 'profile', '--model-repository', '/tmp',
+            '--profile-models', 'model1', '--config-file', '/tmp/my_config.yml',
+            '--checkpoint-directory', '/tmp/my_checkpoints'
+        ]
+        config = self._evaluate_profile_config(args, '')
+        state_manager = AnalyzerStateManager(config, None)
+        analyzer = Analyzer(config, None, state_manager)
+        self.assertEqual(
+            analyzer._get_analyze_command_help_string(),
+            'To analyze the profile results and find the best configurations, '
+            'run `model-analyzer analyze --analysis-models model1 '
+            '--config_file /tmp/my_config.yml --checkpoint-directory '
+            '/tmp/my_checkpoints`')
+
+    def mock_top_n_results(self, model_name=None, n=-1):
+        return [
+            ModelResult(
+                None,
+                ModelConfig(
+                    json_format.ParseDict({'name': 'config1'},
+                                          model_config_pb2.ModelConfig())),
+                None),
+            ModelResult(
+                None,
+                ModelConfig(
+                    json_format.ParseDict({'name': 'config3'},
+                                          model_config_pb2.ModelConfig())),
+                None),
+            ModelResult(
+                None,
+                ModelConfig(
+                    json_format.ParseDict({'name': 'config4'},
+                                          model_config_pb2.ModelConfig())),
+                None)
+        ]
+
+    @patch(
+        'model_analyzer.config.input.config_command_analyze.file_path_validator',
+        lambda _: ConfigStatus(status=CONFIG_PARSER_SUCCESS))
+    @patch(
+        'model_analyzer.config.input.config_command_analyze.ConfigCommandAnalyze._preprocess_and_verify_arguments',
+        lambda _: None)
+    @patch('model_analyzer.result.result_manager.ResultManager.top_n_results',
+           mock_top_n_results)
+    def test_get_report_command_help_string(self):
```

Comment:
can this be moved to the top of the class? Or does it require that some other functions are defined first?

Reply:
https://github.com/triton-inference-server/model_analyzer/pull/250#discussion_r758081557

stuff used in the patch decorator needs to be defined before the python interpreter reaches the patch decorator statement

the stuff above this needs to be above it